### ISSUE TITLE
Execute shellHook on cached runs

### ISSUE THREAD

**Issue by colemickens** [nix-community/nix-direnv#109]
I set PS1 in the shellHook of my `flake.nix` (which I think isn't terrible).

However, after #91 is merged, I think this is ineffective because my shell is basically always cached...

Is there anyway to run `shellHook` in cached hit cases? Or if not, could there be another attribute that is always read out and executed?

**Comment by colemickens** [comment]
Well, it seems like direnv blocks this anyway.

**Comment by Mic92** [comment]
Potentially you could source the cache after `use flake`?

```
eval "$(< "$profile_rc")"
eval $shellHook
```

**Comment by johnae** [comment]
It seems to me as if https://github.com/nix-community/nix-direnv/pull/91 broke https://github.com/numtide/devshell which is unfortunate. Personally I have a simpler take on devshell that works in a similar fashion and that is now broken too. I believe the relevant part of devshell is this:

https://github.com/numtide/devshell/blob/ab14b1a3cb253f58e02f5f849d621292fbf81fad/nix/mkNakedShell.nix#L44

It could be something with my setup that's wrong but it seems I started having problems after ugrading to 1.5.0. @zimbatm could perhaps comment on this as he's the author of devshell.

**Comment by johnae** [comment]
At least I can confirm that deleting this line: https://github.com/nix-community/nix-direnv/blob/d26fc52f8d22540b6bc837120e9c4db7fdec3a4d/direnvrc#L155 makes https://github.com/numtide/devshell work again. What I mean by "not working" is that none of the packages that are supposed to be made available in the environment are.

**Comment by johnae** [comment]
In my case I was able to avoid the issues I experienced by not doing anything "important" in the shellHook so for me this is not a problem anymore.

**Comment by johnae** [comment]
@ryneeverett @Mic92 I believe departing from the behavior of nix develop has made nix-direnv unusable with https://github.com/numtide/devshell. I don't think nix-direnv is doing the right thing here.

**Comment by Mic92** [comment]
This PR enables shellHook for both use_nix and use_flake, even in cached runs
It requires nix2.4 though:

https://github.com/nix-community/nix-direnv/pull/125
I might hold of a release until at least the next nixos version is released and only
publish a release candidate before.

**Comment by lilyball** [comment]
I keep running into this too. I have my own equivalent of numtide/devshell (I didn't know about devshell until now, it sounds like it's a much more comprehensive solution) and the fact that `$shellHook` isn't evaluated on cached runs means I end up with the wrong environment.

@Mic92 I haven't tried the PR, since I don't want to try and shift my config over to pulling patches from GitHub. Given that Nix 2.4 is the stable nix in nixpkgs now, is there a reason not to go ahead and publish a release with this new change? By default a new release won't end up on any existing release channels, so there shouldn't be any concerns there.

**Comment by Mic92** [comment]
> I keep running into this too. I have my own equivalent of numtide/devshell (I didn't know about devshell until now, it sounds like it's a much more comprehensive solution) and the fact that `$shellHook` isn't evaluated on cached runs means I end up with the wrong environment.
> 
> @Mic92 I haven't tried the PR, since I don't want to try and shift my config over to pulling patches from GitHub. Given that Nix 2.4 is the stable nix in nixpkgs now, is there a reason not to go ahead and publish a release with this new change? By default a new release won't end up on any existing release channels, so there shouldn't be any concerns there.

The PR is not ready quite yet.

**Comment by Narice** [comment]
Is there a quickfix for that in the meanwhile?
I tried using the nix-2.4 branch as an overlay to nix-direnv but that didn't work (probably my fault here :smile:, or maybe it should be an overlay on nix-direnv-flakes?)

**Comment by Mic92** [comment]
> Is there a quickfix for that in the meanwhile? I tried using the nix-2.4 branch as an overlay to nix-direnv but that didn't work (probably my fault here smile, or maybe it should be an overlay on nix-direnv-flakes?)

I am using it by sourcing it manually but atm the caching does not work: https://github.com/Mic92/dotfiles/blob/0e30755334eb53e4cbcfe49453b6bf18e3d581c2/home/.direnvrc#L3

**Comment by rraval** [comment]
> Is there a quickfix for that in the meanwhile?

I've added an extra line to my `.envrc` under `use flake`

```
use flake
eval "$shellHook"  # quotes are important here
```

It seems to do the right thing but I've only looked at it superficially and moved on (so I can't comment on whether this is the "right" workaround and what the cons might be).

**Comment by Narice** [comment]
Thanks for both solutions!
That really helps :wink:

**Comment by bbenne10** [comment]
2.0 should handle this case nicely. Please open a new issue if you find some problem with that line. Thank you!
